### PR TITLE
fix: canonicalize equivalent filesystem paths

### PR DESCRIPTION
**Pull request by Nicolas-nwb**
## Summary

- resolve filesystem paths through their nearest existing ancestor for worktree registration and cleanup
- compare global index scope and project ownership using canonical paths without changing catalog identities
- add regressions for symlink aliases, missing leaves, stale file hashes, and legacy branch ownership

## Problem

On macOS, Node can expose temporary paths through `/var/...` while Git records the same worktree under `/private/var/...`. Lexical comparisons can therefore miss a registered temporary worktree and leave stale Git metadata behind.

The same issue affects global indexes opened through symlinked project roots. Once a file is deleted, a one-shot `realpath` cannot resolve the leaf, so file hash state and branch ownership can be misclassified as foreign.

## Approach

Canonicalize the deepest existing ancestor and append any missing path components. Use that equivalence only for registration, scope, and ownership checks. Stored path spelling and catalog identity hashing remain unchanged.

## Testing

- `npx vitest run tests/branch-materialization.test.ts tests/indexer-clear-index.test.ts`
- `npx vitest run tests/automatic-branch-index.test.ts tests/indexer-failed-batches.test.ts`
- `CHOKIDAR_USEPOLLING=1 npm run test:run` (55 files, 1,086 tests)
- `npm run typecheck`
- `npm run lint`
- `npm run build:ts`
- `npm run build:native`
- `git diff --check`

### CODE REVIEW

File: src/indexer/index.ts [hunk 1]
```diff
@@ -2668,7 +2678,14 @@ export class Indexer {
...
   private isFileInCurrentScope(filePath: string, roots: string[]): boolean {
-    return roots.some((root) => isPathWithinRoot(filePath, root));
+    if (roots.some((root) => isPathWithinRoot(filePath, root))) return true;
+    const canonicalFilePath = this.getCanonicalPath(filePath);
+    return roots.some((root) => isPathWithinRoot(canonicalFilePath, root));
+  }
+
+  private isFileInProjectRoot(filePath: string): boolean {
+    if (isPathWithinRoot(filePath, this.projectRoot)) return true;
+    return isPathWithinRoot(this.getCanonicalPath(filePath), this.getCanonicalPath(this.projectRoot));
```
Comment (Helweg) [inline src/indexer/index.ts:2667]: These ownership checks now treat physical and symlink project roots as equivalent, but global catalog identity is still derived from the lexical `projectRoot` in `getBranchCatalogKeyFor`. The two spellings therefore receive different branch catalog keys.

I reproduced this by indexing the physical root and a symlink alias into the same global index. It created two namespaced owners and duplicate chunks for the same file. Deleting through the physical spelling left alias-owned chunks and branch rows behind.

Could the project identity use the canonical root, or could an explicit alias migration/ownership mapping consolidate existing lexical identities? Please add a regression covering physical-root indexing followed by alias-root indexing and cleanup.

File: src/git/branch-materialization.ts [hunk 2]
```diff
@@ -421,13 +421,38 @@ async function pathExists(targetPath: string): Promise<boolean> {
...
+  while (true) {
+    try {
+      return path.join(await fsPromises.realpath(candidate), ...missingParts);
+    } catch (error) {
+      const code = (error as NodeJS.ErrnoException).code;
+      if (code !== "ENOENT" && code !== "ENOTDIR") throw error;
+
+      const parent = path.dirname(candidate);
+      if (parent === candidate) return resolved;
+      missingParts.unshift(path.basename(candidate));
```
Comment (Helweg) [inline src/git/branch-materialization.ts]: Appending missing components with their original casing does not produce a stable identity on case-insensitive filesystems. After the leaf disappears, equivalent paths such as `Missing/File.ts` and `missing/file.ts` canonicalize to different strings even on default macOS and Windows volumes.

Because these values are later compared using exact equality, a stale worktree registration can still be missed. Could missing suffixes be normalized according to the platform/filesystem semantics, with macOS and Windows regression coverage?

**Review by Helweg — CHANGES_REQUESTED** [review]
Thanks for addressing path equivalence across missing leaves and symlink aliases. The focused implementation is promising and its existing validation passes, but two identity cases remain before this is safe to merge.

File: src/git/branch-materialization.ts (hunk 2 above)
Comment (Nicolas-nwb) [reply src/git/branch-materialization.ts]: Addressed in `b4ec6b2`.

The nearest-existing-ancestor logic now lives in the shared `canonicalizePathForComparison` helper. Missing suffix components are ASCII-folded only when the hosting filesystem is positively detected as case-insensitive; case-sensitive and unknown filesystems preserve the original spelling. Empty Windows directories use a transient exclusive probe that is removed immediately.

The new tests cover confirmed insensitive, sensitive, unknown, populated-host, and empty-directory behavior. A focused CI matrix now runs the path semantics test on Ubuntu, macOS, and Windows, and all three jobs plus the main CI passed.

File: src/indexer/index.ts (hunk 1 above)
Comment (Nicolas-nwb) [reply src/indexer/index.ts:2667]: Addressed in `b4ec6b2`.

Global project identity is now captured from the canonicalized project root and reused for branch catalogs and all project-scoped global metadata. Equivalent namespaced owners that reference project-local chunks or symbols are also treated as current-project ownership during cleanup, so prior lexical aliases do not keep data alive.

I added a regression that indexes the physical root, then a symlink alias, asserts one catalog with no duplicate chunks, seeds a lexical alias owner, and clears it through the physical root.

**Comment by Helweg** [comment]
Thanks for the updates. The two previously requested identity cases appear to be addressed, and the focused tests and CI matrix are passing.

I found one remaining edge case in `foldMissingPathComponent`. It only lowercases ASCII characters matching `[A-Z]`, while case-insensitive filesystems can also equate cased Unicode characters. This affects, for example:

- Danish: `Æ/æ`, `Ø/ø`, `Å/å`
- French/Italian: `È/è`, `É/é`, `À/à`
- German: `Ä/ä`, `Ö/ö`, `Ü/ü`
- Greek and other scripts with uppercase/lowercase forms

I reproduced this even with `isCaseInsensitive: () => true`: equivalent missing paths such as `Årsrapport.ts` and `årsrapport.ts` still produce different canonical identities.

Could the missing components use Unicode-aware case folding, with regression coverage for non-ASCII filenames and directory names?

**Comment by Nicolas-nwb** [comment]
Thanks for sticking with this and for the clear reproduction. Addressed in `f58fbb7`.

You were right: ASCII-only folding left non-ASCII case pairs with different identities.

I added red regressions for accented Latin directory and filename components, plus context-independent Greek sigma handling, where `toLowerCase()` is not sufficient. Missing components now use Unicode Default Full Case Folding, but only after the filesystem is positively confirmed case-insensitive; case-sensitive and unknown filesystems remain unchanged.

The folding implementation is pinned, has no transitive dependencies, and is bundled for both ESM and CommonJS builds. The focused tests, full test suite, typecheck, lint, build, and production dependency audit all pass. An independent review found no issues.

**Review by Helweg — APPROVED** [review]
Re-reviewed updated head f58fbb7. The remaining Unicode case-folding issue is addressed, the affected regression suites and build checks pass, and I found no remaining actionable issues.
